plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSink.java

```diff
@@ -214,40 +211,85 @@ private void destroy()
         @GuardedBy("this")
         private SliceOutput currentBuffer;
-
-        public BufferedStorageWriter(ExchangeStorageWriter storageWriter, BufferPool bufferPool, AtomicReference<Throwable> failure)
+        @GuardedBy("this")
+        private final List<ExchangeStorageWriter> writers = new ArrayList<>();
+
+        public BufferedStorageWriter(
+                FileSystemExchangeStorage exchangeStorage,
+                URI outputDirectory,
+                Optional<SecretKey> secretKey,
+                int partitionId,
+                BufferPool bufferPool,
+                AtomicReference<Throwable> failure,
+                long exchangeSinkFileTargetSize)
         {
-            this.storageWriter = requireNonNull(storageWriter, "storageWriter is null");
+            this.exchangeStorage = requireNonNull(exchangeStorage, "exchangeStorage is null");
+            this.outputDirectory = requireNonNull(outputDirectory, "outputDirectory is null");
+            this.secretKey = requireNonNull(secretKey, "secretKey is null");
+            this.partitionId = partitionId;
             this.bufferPool = requireNonNull(bufferPool, "bufferPool is null");
             this.failure = requireNonNull(failure, "failure is null");
+            this.exchangeSinkFileTargetSize = exchangeSinkFileTargetSize;
+
+            updateCurrentWriter();
         }
 
         public synchronized void write(Slice data)
         {
+            if (currentFileSize >= exchangeSinkFileTargetSize) {
+                flushIfNeeded(true);
+                updateCurrentWriter();
+                currentFileSize = 0;
+                currentBuffer = null;
+            }
+
             writeInternal(Slices.wrappedIntArray(data.length()));
             writeInternal(data);
+
+            currentFileSize += Integer.BYTES + data.length();
         }
 
         public synchronized ListenableFuture<Void> finish()
         {
             flushIfNeeded(true);
-            return storageWriter.finish();
+            if (writers.size() == 1) {
+                return currentWriter.finish();
+            }
+            return asVoid(Futures.allAsList(writers.stream().map(ExchangeStorageWriter::finish).collect(toImmutableList())));
         }
 
         public synchronized ListenableFuture<Void> abort()
         {
-            return storageWriter.abort();
+            if (writers.size() == 1) {
+                return currentWriter.abort();
+            }
+            return asVoid(Futures.allAsList(writers.stream().map(ExchangeStorageWriter::abort).collect(toImmutableList())));
         }
 
         public synchronized long getRetainedSize()
         {
-            return INSTANCE_SIZE + storageWriter.getRetainedSize();
+            return INSTANCE_SIZE + estimatedSizeOf(writers, ExchangeStorageWriter::getRetainedSize);
+        }
+
+        private void updateCurrentWriter()
+        {
+            currentWriter = exchangeStorage.createExchangeStorageWriter(
```

Comment:
We have to make sure no more writers are accidentally created after `abort`.

Reply:
Good catch. Added a variable `closed` to protect against such cases.